### ISSUE TITLE
Finalizer does not seem to be called...

### ISSUE THREAD

**Issue by errm** [ledbettj/systemd-journal#73]
Hi there me again,

I have [this issue](https://github.com/reevoo/fluent-plugin-systemd/issues/34) raised on fluent-plugin-systemd , it seems that on rotation we keep journal files open, even after they are deleted.

When we get `:invalidate` from `Systemd::Journal#wait` we new up a whole new instance of `Systemd::Journal` here https://github.com/reevoo/fluent-plugin-systemd/blob/master/lib/fluent/plugin/in_systemd.rb#L64

As I understand the finalizer [here](https://github.com/ledbettj/systemd-journal/blob/master/lib/systemd/journal.rb#L276) should be invoked at some point after this when the object is garbage collected...

I would appreciate it if you have some pointers to track down what is going wrong here?

Also I was thinking it might be good to just expose a method to manually call sd_journal_close.  I don't know what your thoughts on this are?

**Comment by ledbettj** [comment]
Interesting, thanks for the report.  When ruby decides to run finalizers is not super clear to me.  The number one reason why a finalizer might not get run is that there is still a reference to that object somewhere else in the memory.  Looking at your code that's probably not the case.

Running some simple tests (with a `puts` in the finalizer) I can see:

```
[1] pry(main)> 100.times { s = Systemd::Journal.new }
finalizing #<FFI::Pointer address=0x0000dcf5249470>
finalizing #<FFI::Pointer address=0x0000dcf5249950>
finalizing #<FFI::Pointer address=0x0000dcf5248f70>
finalizing #<FFI::Pointer address=0x0000dcf5249120>
finalizing #<FFI::Pointer address=0x0000dcf5248d40>
... (snip)
=> 100
[2] pry(main)> x = Systemd::Journal.new
=> #<Systemd::Journal:0x0000006e7a56244c target: "journal:local", flags: 0x00000000>
[3] pry(main)> x = Systemd::Journal.new
=> #<Systemd::Journal:0x0000006e7a57a8e4 target: "journal:local", flags: 0x00000000>
[15] pry(main)> GC.start
=> nil
[16] pry(main)> exit
finalizing #<FFI::Pointer address=0x0000dcf4dab200>
finalizing #<FFI::Pointer address=0x0000dcf50c02f0>
```

In the first case with the journal referenced inside a loop, it's freed immediately after each iteration, probably because ruby has identified it as a short-lived variable.

In the second case, it's not cleaned up, even with an explicit GC.start until process exit.  This is definitely not desirable when you are closing files, since you can end up running out of open file descriptors.

I agree with you that a good solution would be to default to this automatic management, but allowing the user to specify that they will manually close the journal.

Thanks!

**Comment by ledbettj** [comment]
I'll follow up here when I've pushed a new version with functionality from #74 !

**Comment by ledbettj** [comment]
Hopefully fixed in v1.3.0.

**Comment by errm** [comment]
Thanks so much @ledbettj looks great
